File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel


class MoodPoint(BaseModel):
    date: str
    average_mood: float
    entries_count: int


class JournalAnalyticsSummary(BaseModel):
    total_entries: int
    average_mood: float | None = None
    min_mood: int | None = None
    max_mood: int | None = None


class JournalRecentEntry(BaseModel):
    id: int
    title: str
    mood_score: int
    is_private: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class JournalStreakResponse(BaseModel):
    current_streak: int
    longest_streak: int
    total_active_days: int


class JournalAnalyticsResponse(BaseModel):
    summary: JournalAnalyticsSummary
    mood_history: list[MoodPoint]


class MoodAnalyticsPoint(BaseModel):
    period_start: str
    period_end: str
    label: str
    average_mood: float | None = None
    entries_count: int


class TopEmotionPoint(BaseModel):
    emotion_label: str
    count: int
    percentage: float


class JournalingFrequency(BaseModel):
    total_days: int
    active_days: int
    entries_count: int
    entries_per_week: float
    average_entries_per_active_day: float
    consistency_percentage: float


class StreakCalendarDay(BaseModel):
    date: str
    has_entry: bool
    entries_count: int
    average_mood: float | None = None
    streak_day: int


class CorrelationMetric(BaseModel):
    metric: str
    coefficient: float | None = None
    strength: str
    interpretation: str


class EmotionHeatmapDay(BaseModel):
    date: str
    dominant_emotion: str | None = None
    entries_count: int
    average_mood: float | None = None
    intensity: float


class MoodAnalyticsInsight(BaseModel):
    title: str
    description: str


class MoodAnalyticsResponse(BaseModel):
    period: str
    granularity: str
    start_date: str
    end_date: str
    summary: JournalAnalyticsSummary
    mood_history: list[MoodAnalyticsPoint]
    top_emotions: list[TopEmotionPoint]
    journaling_frequency: JournalingFrequency
    streak: JournalStreakResponse
    streak_calendar: list[StreakCalendarDay]
    correlations: list[CorrelationMetric]
    emotion_heatmap: list[EmotionHeatmapDay]
    insights: list[MoodAnalyticsInsight]

File: mobile/src/api/analytics.ts
import { apiFetch } from './client';

export type MoodAnalyticsPoint = {
  period_start: string;
  period_end: string;
  label: string;
  average_mood: number | null;
  entries_count: number;
};

export type TopEmotionPoint = {
  emotion_label: string;
  count: number;
  percentage: number;
};

export type JournalingFrequency = {
  total_days: number;
  active_days: number;
  entries_count: number;
  entries_per_week: number;
  average_entries_per_active_day: number;
  consistency_percentage: number;
};

export type CorrelationMetric = {
  metric: string;
  coefficient: number | null;
  strength: string;
  interpretation: string;
};

export type EmotionHeatmapDay = {
  date: string;
  dominant_emotion: string | null;
  entries_count: number;
  average_mood: number | null;
  intensity: number;
};

export type MoodAnalyticsInsight = {
  title: string;
  description: string;
};

export type MoodAnalytics = {
  period: string;
  granularity: string;
  start_date: string;
  end_date: string;
  summary: {
    total_entries: number;
    average_mood: number | null;
    min_mood: number | null;
    max_mood: number | null;
  };
  mood_history: MoodAnalyticsPoint[];
  top_emotions: TopEmotionPoint[];
  journaling_frequency: JournalingFrequency;
  streak: {
    current_streak: number;
    longest_streak: number;
    total_active_days: number;
  };
  streak_calendar: Array<{
    date: string;
    has_entry: boolean;
    entries_count: number;
    average_mood: number | null;
    streak_day: number;
  }>;
  correlations: CorrelationMetric[];
  emotion_heatmap: EmotionHeatmapDay[];
  insights: MoodAnalyticsInsight[];
};

export function getMoodAnalytics(period = '30d', granularity = 'day') {
  const query = `?period=${encodeURIComponent(period)}&granularity=${encodeURIComponent(granularity)}`;
  return apiFetch<MoodAnalytics>(`/analytics/mood${query}`, { method: 'GET', auth: true });
}
